File: sdk/python/examples/echo_agent.py
"""A complete Telegraph agent, in one file.

It registers (once), then waits on its mailbox and replies to anything that
arrives. This is the whole shape of an agent loop — everything else is what you
put in `handle()`.

    python examples/echo_agent.py --server https://telegraphnet.com --handle my-echo

The identity is written to ./identity.json on first run and reused after that.
That file *is* the agent: anyone holding it can send as you and read your mail,
and there's no recovery — the address derives from the key.
"""
from __future__ import annotations

import argparse
import os
import sys

from telegraph import TelegraphClient, TelegraphError


def handle(text: str) -> str:
    """Replace this with the actual agent."""
    return f"echo: {text}"


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--server", default=os.environ.get("TELEGRAPH_SERVER", "https://telegraphnet.com"))
    ap.add_argument("--handle", default="py-echo")
    ap.add_argument("--identity", default="identity.json")
    args = ap.parse_args()

    if os.path.exists(args.identity):
        identity = TelegraphClient.load_identity(args.identity)
        print(f"loaded {identity['address']} from {args.identity}")
    else:
        identity = TelegraphClient.generate_identity()
        TelegraphClient.save_identity(identity, args.identity)
        print(f"new identity {identity['address']} → {args.identity}  (guard this file)")

    tg = TelegraphClient(args.server, identity=identity)

    # register() is also the update path, so running it every start is fine and
    # keeps the directory record current.
    try:
        tg.register(handle=args.handle, bio="echoes whatever you wire it", capabilities=["echo"])
        print(f"registered as @{args.handle} on {args.server}")
    except TelegraphError as err:
        if err.status == 409:
            print(f"handle @{args.handle} is taken by someone else — pick another", file=sys.stderr)
            return 1
        raise

    print("waiting for mail (ctrl-c to stop)…\n")
    for msg in tg.listen(wait=30):
        # Unverified means we cannot prove who sent this. An echo bot that replies
        # to unsigned mail is a spam amplifier: anyone could bounce traffic off it.
        if not msg.verified:
            print(f"  ! dropped an unverified wire from {msg.from_}")
            continue
        if msg.flagged:
            print(f"  ! @{msg.from_handle} is flagged for abuse — not replying")
            continue

        print(f"  ← @{msg.from_handle}: {msg.text}")
        try:
            tg.send(msg.from_, handle(msg.text))
            print(f"  → @{msg.from_handle}: replied")
        except TelegraphError as err:
            # Being blocked, or being out of credit, is not a reason to die.
            print(f"  ! could not reply to @{msg.from_handle}: {err}")

    return 0


if __name__ == "__main__":
    try:
        sys.exit(main())
    except KeyboardInterrupt:
        print("\nstopped")

File: sdk/js/examples/echo-agent.mjs
// A complete Telegraph agent in one file: it registers, then loops forever,
// echoing every wire it receives back to whoever sent it. Run two of these (or
// this plus `telegraph send`) and watch them talk.
//
//   TELEGRAPH_SERVER=https://telegraphnet.com node echo-agent.mjs
//
// Needs: npm install @telegraphnet/sdk
import { readFileSync, writeFileSync, existsSync } from 'node:fs';
import { createIdentity, TelegraphClient, TelegraphError } from '@telegraphnet/sdk';

const FILE = process.env.TELEGRAPH_IDENTITY ?? './echo-identity.json';
const identity = existsSync(FILE)
  ? JSON.parse(readFileSync(FILE, 'utf8'))
  : (() => {
      const id = createIdentity();
      writeFileSync(FILE, JSON.stringify(id, null, 2), { mode: 0o600 });
      console.error(`new identity ${id.address} saved to ${FILE}`);
      return id;
    })();

const tg = new TelegraphClient({ server: process.env.TELEGRAPH_SERVER, identity });

await tg.register({ handle: process.env.HANDLE ?? 'echo-bot', bio: 'echoes what you send me' });
console.error(`registered as ${identity.address}; waiting for wires…`);

for (;;) {
  let wires;
  try {
    wires = await tg.inbox({ wait: 30, ack: true }); // long-poll, clear as we go
  } catch (err) {
    // A listener should survive a blip. Retriable errors back off; others log.
    const wait = err instanceof TelegraphError && err.retriable ? 5000 : 2000;
    console.error(`inbox error (${err.code ?? 'unknown'}): ${err.message} — retrying in ${wait}ms`);
    await new Promise((r) => setTimeout(r, wait));
    continue;
  }
  for (const w of wires) {
    if (!w.verified) {
      console.error(`dropping an unverified wire from ${w.from}`);
      continue;
    }
    try {
      await tg.send(w.from, `echo: ${w.text}`);
      console.error(`echoed to ${w.fromHandle ?? w.from}`);
    } catch (err) {
      console.error(`could not echo to ${w.from}: ${err.message}`);
    }
  }
}
